enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,747 @@
+
+[[customizations.files]]
+path = "/etc/microshift/config.yaml"
+data = """
+kubelet:
+  cpuManagerPolicy: static
+  memoryManagerPolicy: Static
+"""
+
+[[customizations.files]]
+path = "/etc/microshift/microshift-tuned.yaml"
+data = """
+reboot_after_apply: True
+profile: microshift-baseline
+"""
+```
+
+
+##### bootc
+
+1. User creates Containerfile that:
+   - (optional) installs `kernel-rt`
+   - installs `microshift-low-latency.rpm`
+   - enables `microshift-tuned.service`
+   - adds following configs
+     - `/etc/tuned/microshift-baseline-variables.conf`
+     - `/etc/microshift/config.yaml` to configure Kubelet
+     - `/etc/microshift/microshift-tuned.yaml` to configure `microshift-tuned.service`
+1. User builds the blueprint
+1. User deploys the commit / installs the system.
+1. System boots - rest is just like in OSTree flow
+
+Example Containerfile:
+
+```
+FROM registry.redhat.io/rhel9/rhel-bootc:9.4
+
+# ... MicroShift installation ...
+
+RUN dnf install kernel-rt microshift-tuned
+COPY microshift-baseline-variables.conf /etc/tuned/microshift-low-latency-variables.conf
+COPY microshift-config.yaml             /etc/microshift/config.yaml
+COPY microshift-tuned.yaml              /etc/microshift/microshift-tuned.yaml
+
+RUN systemctl enable microshift-tuned.service
+```
+
+##### RPM
+
+1. User installs `microshift-low-latency` RPM.
+1. User creates following configs:
+   - `/etc/tuned/microshift-baseline-variables.conf`
+   - `/etc/microshift/config.yaml` to configure Kubelet
+1. Development environment
+   - User runs `sudo tuned-adm profile microshift-baseline` to enable the profile.
+   - User reboots the host to make changes to kernel arguments active.
+   - Host boots again, everything for low latency is in place,
+   - User starts/enables `microshift.service`
+1. Production environment
+   - User creates `/etc/microshift/microshift-tuned.yaml` to configure `microshift-tuned.service`
+   - User enables `microshift.service`
+   - User enables and starts `microshift-tuned.service` which:
+     - Compares active profile with requested profile
+     - If requested profile is already active:
+       - Compare checksum of requested profile with cached checksum.
+       - If checksums are the same - exit.
+     - Apply requested profile
+     - Calculate checksum of the profile and the variables file and save it
+     - If `reboot_after_apply` is True, then reboot the host
+   - Host is rebooted: MicroShift starts because it was enabled
+   - Host doesn't need reboot:
+     - User starts `microshift.service`
+
+
+#### Preparing low latency workload
+
+- Setting `.spec.runtimeClassName: microshift-low-latency` in Pod spec.
+- Setting Pod's memory limit and memory request to the same value, and
+  setting CPU limit and CPU request to the same value to ensure Pod has guaranteed QoS class.
+- Use annotations to get desired behavior
+  (unless link to a documentation is present, these annotations only take two values: enabled and disabled):
```

Comment:
these annotations only take two values: "enable" and "disable".

Reply:
Here's the code in cri-o where the value of the load balancing annotation is evaluated:

```
const (
	annotationTrue       = "true"
	annotationDisable    = "disable"
	annotationEnable     = "enable"
)

func shouldCPULoadBalancingBeDisabled(annotations fields.Set) bool {
	if annotations[crioannotations.CPULoadBalancingAnnotation] == annotationTrue {
		log.Warnf(context.TODO(), annotationValueDeprecationWarning(crioannotations.CPULoadBalancingAnnotation))
	}

	return annotations[crioannotations.CPULoadBalancingAnnotation] == annotationTrue ||
		annotations[crioannotations.CPULoadBalancingAnnotation] == annotationDisable
}
```
I was pointing out that saying "(unless link to a documentation is present, these annotations only take two values: `enabled` and `disabled`):" is inaccurate - it should read "(unless link to a documentation is present, these annotations only take two values: `enable` and `disable`):" Specifying `cpu-load-balancing.crio.io: "disabled" results in the opposite behavior as`cpu-load-balancing.crio.io: "disable"`.